[3D] Fix combination of rotation, combination and zoom

ptitjano opened the pull request:
## Description

Different kind of operations can happen while clicking on the mouse: zoom, rotation or translation. However, combining these different operations while keeping the mouse pressed does not work. This PR fixes those issues.


### Rotation + Zoom In + Rotation

#### Current behavior

The first rotation and the zoom in work but  at the second rotation, the view is reset as if the zoom never happened

[rotate-zoom_broken.webm](https://github.com/qgis/QGIS/assets/497207/9423ce1e-675a-4533-a43a-c6da42957fa3)

#### New behavior

[rotate-zoom_ok.webm](https://github.com/qgis/QGIS/assets/497207/5d0a4e07-ac07-43e3-b17a-61dc8a6187a9)

### Translation + Rotation

#### Current behavior

Once the rotation starts, the camera parameters are broken. 

[drag-rotate_broken.webm](https://github.com/qgis/QGIS/assets/497207/17ee367b-7108-43db-9666-ce366afdf697)

#### New behavior

[drag-rotate_ok.webm](https://github.com/qgis/QGIS/assets/497207/16fcd7b0-98eb-40a1-b256-f82afe510875)


cc @benoitdm-oslandia @soaubier

ptitjano committed:
qgscameracontroller: Fix combination of rotation and zoom with the mouse

While pressing the left button of the mouse, some sequences do not
work properly.
For example, : rotate around a clicked point, zoom-in and then rotate
again. The last rotation does not work because a new rotation
center (mRotationCenter) has not been computed. This is because the
`mRotationCenterCalculated` flag has not been updated.

This issue is fixed by calling
`QgsCameraController::setMouseParameters` accordingly in
`QgsCameraController::onPositionChangedTerrainNavigation`.

ptitjano committed:
qgscameracontroller: Remove unused attribute mPressedButton

ptitjano committed:
qgscameracontroller: Remove unused attribute mMousePressed

ptitjano committed:
qgscameracontroller: Move mouse parameters logic in a unique function

Different kind of operations can happen while clicking on the mouse:
zoom, rotation or translation. A lot of parameters need to be updated
to keep track of the current operation.

This commit only moves all this logic in the
`QgsCameraController::setMouseParameters` function. This makes it
easier to understand the code base and this will make it easier to
fix some existing issues in the next commits.

ptitjano committed:
qgscameracontroller: Do not set both translate and rotate on mousepressed

It cannot happen a translation and a rotation at the same time.

ptitjano committed:
qgscameracontroller: Use updateCameraFromPose in setCameraPose

ptitjano committed:
qgscameracontroller: Use only one camerabefore

These different cameras are never used at the same time.

ptitjano committed:
qgscameracontroller: Fix combination of translate and rotate with the mouse

While pressing the left button of the mouse, some sequences do not
work properly.
For example, translate, and then rotate on the clicked point. The
rotation does not work because the initial rotation angles and the
rotation center are not computed when the rotation starts.

This issue is fixed by adding a clickpoint parameter to
`QgsCameraController::setMouseParameters`. This allows to set the
rotation center to the mouse position when the rotation starts.
However, the new `mClickPoint` must not be updated if the two
consecutive operations are not a rotation or a translation because
that would break a sequence if there is a zoom operation between a
rotation or a translation.

nyalldawson commented on src/3d/qgscameracontroller.h:
```diff
@@ -225,6 +225,16 @@ class _3D_EXPORT QgsCameraController : public QObject
     //! Returns a pointer to the scene's engine's window or nullptr if engine is QgsOffscreen3DEngine
     QWindow *window() const;
 
+    enum MouseOperation
```
```suggestion
    enum class MouseOperation
```

ptitjano replied:
Done.

ptitjano committed:
qgscameracontroller: Fix a clang-tidy warning

warning: narrowing conversion from 'double' to 'float'
[bugprone-narrowing-conversions]

nyalldawson commented on src/3d/qgscameracontroller.cpp:
```diff
@@ -623,64 +611,34 @@ void QgsCameraController::onWheel( Qt3DInput::QWheelEvent *wheel )
 
-  mDragPointCalculated = false;
-  mRotationCenterCalculated = false;
+  mClickPoint = QPoint( 0, 0 );
```
```suggestion
  mClickPoint = QPoint();
```

nyalldawson commented on src/3d/qgscameracontroller.cpp:
```diff
@@ -1086,6 +1044,48 @@ void QgsCameraController::depthBufferCaptured( const QImage &depthImage )
+  // Indeed, if the sequence such as rotation -> zoom -> rotation updating mClickPoint on
+  // the click point does not need to be updated because the relative mouse position is kept
+  // during a zoom operation
+  if ( mClickPoint == QPoint( 0, 0 ) ||
```
```suggestion
  if ( mClickPoint.isNull() ||
```

nyalldawson commented on src/3d/qgscameracontroller.h:
```diff
@@ -225,6 +225,16 @@ class _3D_EXPORT QgsCameraController : public QObject
+      Zoom
+    };
+
+    void setMouseParameters( const MouseOperation &newOperation, const QPoint &clickPoint = QPoint( 0, 0 ) );
```
```suggestion
    void setMouseParameters( const MouseOperation &newOperation, const QPoint &clickPoint = QPoint() );
```

nyalldawson commented on src/3d/qgscameracontroller.h:
```diff
@@ -294,28 +304,25 @@ class _3D_EXPORT QgsCameraController : public QObject
-    Qt3DInput::QMouseEvent::Buttons mPressedButton = Qt3DInput::QMouseEvent::Buttons::NoButton;
+
+    //! click point for a rotation or a translation
+    QPoint mClickPoint = QPoint( 0, 0 );
```
```suggestion
    QPoint mClickPoint;
```

nyalldawson committed:
Apply suggestions from code review

nyalldawson commented:
@ptitjano Great work, this was an annoying bug! 

I did spot a regression here though -- you can no longer zoom in and out of scenes by holding the right mouse button.

It's ntot caused by your changes, but I've also noticed there's issues if you hold left button and pan the scene, then while the button is held press ctrl to start rotating and then release ctrl. Is a similar fix possible to address this one?  

(Things also go wonky when you click and hold left button and then wheel to zoom at the same time, but again, that's not caused by your fixes! :rofl: )

ptitjano commented:
> @ptitjano Great work, this was an annoying bug!
> 
> I did spot a regression here though -- you can no longer zoom in and out of scenes by holding the right mouse button.
> 
> It's ntot caused by your changes, but I've also noticed there's issues if you hold left button and pan the scene, then while the button is held press ctrl to start rotating and then release ctrl. Is a similar fix possible to address this one?
> 
> (Things also go wonky when you click and hold left button and then wheel to zoom at the same time, but again, that's not caused by your fixes! 🤣 )

Thanks for testing.
All those issues should be fixed by https://github.com/qgis/QGIS/pull/55382
